src/util/misc.rs

```diff
@@ -161,18 +168,15 @@ mod message_signing {
             secp_ctx: &secp256k1::Secp256k1<C>,
             address: &Address,
             msg_hash: sha256d::Hash
-        ) -> Result<bool, secp256k1::Error> {
+        ) -> Result<bool, MessageSignatureError> {
             let pubkey = self.recover_pubkey(secp_ctx, msg_hash)?;
-            Ok(match address.address_type() {
+            match address.address_type() {
                 Some(AddressType::P2pkh) => {
-                    *address == Address::p2pkh(&pubkey, address.network)
+                    Ok(*address == Address::p2pkh(&pubkey, address.network))
                 }
-                Some(AddressType::P2sh) => false,
-                Some(AddressType::P2wpkh) => false,
-                Some(AddressType::P2wsh) => false,
-                Some(AddressType::P2tr) => false,
-                None => false,
-            })
+                Some(address_type) => Err(MessageSignatureError::UnsupportedAddressType(address_type)),
```

Comment:
`Some(ty @ AddressType::P2sh) | Some(ty @ AddressType::P2wpkh) | Some(ty @ AddressType::P2wsh) | Some(ty @ AddressType::P2tr)` would be more refactor-proof but I guess it's unlikely a new address will be introduced together with a signature scheme?

Reply:
If we base the current behaviour off of this comment: https://github.com/rust-bitcoin/rust-bitcoin/pull/684#issuecomment-1011314570
> There is no signmessage standard for non-pkh address types.

Then all other address types would be unsupported and we wouldn't need to explicitly list each. I can go either way. No strong opinions from my end.

Reply:
Yes, it is very unlikely that any new signature schemes are going to show up which support onlly specific address types.